### PR TITLE
[FSDP2] Added CPU offloading

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124513
* __->__ #120256
* #124319

#### Overview
This PR adds CPU offloading via the `offload_policy: OffloadPolicy` argument.
- We incur one H2D copy for each parameter before all-gather.
- We incur one D2H copy for each gradient after reduce-scatter.
- We run optimizer on CPU.

#### Example (Mixed Precision and CPU Offloading)
This example uses a small 125M numel model, which is not too representative. We can try to run with a larger model like Llama-7B. However, since the current optimizer step is already too slow, we may want to patch a faster CPU optimizer.

Forward
Screenshot 2024-02-21 at 10 36 29 AM
Screenshot 2024-02-21 at 10 39 12 AM

Backward
Screenshot 2024-02-21 at 10 37 47 AM
Screenshot 2024-02-21 at 10 38 44 AM

Overall CPU (CPU optimizer step dominates)
Screenshot 2024-02-21 at 10 39 47 AM

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

**Comment by awgu** [comment]
cc: @msaroufim 

With this PR, per-parameter FSDP (`fully_shard`) has preliminary support for CPU offloading.
- We do not have meta-device init yet. One approach is to replicate existing FSDP's approach to meta-device init, where the unsharded module is materialized on GPU and immediately sharded, requiring either `reset_parameters()` or a user-defined function, but I want to defer this since we can have a cleaner way once we land some `DTensor` infra. See https://github.com/pytorch/pytorch/issues/120003.
- For now, the model can be initialized on CPU. We need to test/flesh out the checkpoint load flow, but this is orthogonal to CPU offloading. We can get some help from @fegin @wz337 @LucasLLC there.
- We shard the model with FSDP. For a transformer model, this effectively means:
  ```
  for module in model.modules():
    if isinstance(module, TransformerBlock):
      fully_shard(module, offload_policy=OffloadPolicy("cpu"))
  fully_shard(model, offload_policy=OffloadPolicy("cpu"))
  ```
- Each `TransformerBlock` forms one parameter group, and the root module (`model`) forms a final group (of the embedding and output projection). Each group's parameters are communicated together via all-gather (and similarly its gradients for reduce-scatter). If CPU offloading, then the group's parameters are copied H2D at the same point, and its gradients are copied D2H at the same point. If you want smaller groups, you can consider applying `fully_shard` at a finer granularity, e.g. on the `Attention` and `Feedforward` within the `TransfomerBlock`.
- During initialization, the current implementation will move the group's parameters to GPU, shard, and offload them back to CPU. This should not increase peak memory since those parameters would be moved to GPU together anyway for forward/backward computation.
- FSDP will implicitly 'prefetch'. If CPU offloading, then the next group's parameters are copied H2D during the previous group's forward. In backward, the gradient D2H copy will be async as long as we are not accumulating gradients (aka microbatching). The PR description shows some toy traces to exemplify the overlap.

One caveat I was thinking about is that FSDP assumes distributed is available and will try to initialize a process group if not passed one. For single-GPU setups, perhaps this assumption is a hindrance. As such, I am not sure whether it makes sense to couple single-GPU CPU offloading with FSDP's CPU offloading. (Though, having a fragmented story would also be a downside.)

**Comment by weifengpy** [comment]
I am testing this PR on FSDP + llama-7B

File: test/distributed/_composable/fsdp/test_fully_shard_memory.py
Comment (Skylion007) [inline test/distributed/_composable/fsdp/test_fully_shard_memory.py:132]: Should we have some episilon constant here?

Comment (awgu) [reply test/distributed/_composable/fsdp/test_fully_shard_memory.py:132]: Sorry, to clarify, what would the epsilon be for?

**Comment by awgu** [comment]
I am thinking a bit on the API design. Should we expose a way for users to apply these kinds of policies (e.g. mixed precision, CPU offloading) at tensor level? If so, how?

**Comment by weifengpy** [comment]
> I am thinking a bit on the API design. Should we expose a way for users to apply these kinds of policies (e.g. mixed precision, CPU offloading) at tensor level? If so, how?

is there some chance to decouple cpu offloading as indepedent pre/post-forward/backward hooks? so offloading can work for non-distributed and distributed ? 

for forward, H2D hook -> all-gather hook 
for backward, reduce-scatter hook -> D2H hook

**Comment by awgu** [comment]
> > I am thinking a bit on the API design. Should we expose a way for users to apply these kinds of policies (e.g. mixed precision, CPU offloading) at tensor level? If so, how?
> 
> is there some chance to decouple cpu offloading as indepedent pre/post-forward/backward hooks? so offloading can work for non-distributed and distributed ?
> 
> for forward, H2D hook -> all-gather hook for backward, reduce-scatter hook -> D2H hook

To summarize, CPU offloading for FSDP looks like four parts, all at the tensor level (as oppose to module level):
- Post-sharded-parameter init: offload the sharded parameter's data (padded) to CPU pinned memory and initialize a CPU sharded gradient in pinned memory
- Pre-all-gather: copy the sharded parameter from H2D
- Post-reduce-scatter: copy the sharded gradient from D2H, optionally as non-blocking if not accumulating gradients, and record a CUDA event
- Finalize backward (end of backward): block CPU on the recorded CUDA event

The CUDA event part is only necessary because FSDP uses a separate reduce-scatter stream.

I am not sure how useful it is to factor out these parts to be usable by single-GPU (non-FSDP) because I think the majority of the heavy lifting is done not by these additional four parts but rather by FSDP's _module-level_ forward/backward hook logic to call the pre-all-gather H2D copy and post-reduce-scatter D2H copy at a good time.

**Comment by awgu** [comment]
If we want to allow users to offload just some of the parameters in a parameter group, then we could follow the tensor parallel API's approach by requiring a `Dict[str, ...]` that maps from fully-qualified name to policy.

```
offload_policy = {
  "feed_forward.w1": OffloadPolicy("cpu"),
  "feed_forward.w2": OffloadPolicy("cpu"),
  "feed_forward.w3": OffloadPolicy("cpu"), 
}
```

A few issues:
- The FQNs need to be prefixed from the `module` passed to `fully_shard(module)`, which may change with different "wrapping" policy.
- This approach might make sense for CPU offloading because it is purely a tensor-level construct. However, for something like mixed precision, we confound module-level constructs like casting the forward inputs/outputs to a dtype with tensor-level constructs like the parameter/reduce dtypes. It is less clear how this `Dict[str, ...]` approach would work then. cc: @wanchaol

**Comment by msaroufim** [comment]
> . For single-GPU setups, perhaps this assumption is a hindrance. As such, I am not sure whether it makes sense to couple single-GPU CPU offloading with FSDP's CPU offloading. (Though, having a fragmented story would also be a downside.)

> I am not sure how useful it is to factor out these parts to be usable by single-GPU (non-FSDP) because I think the majority of the heavy lifting is done not by these additional four parts but rather by FSDP's module-level forward/backward hook logic to call the pre-all-gather H2D copy and post-reduce-scatter D2H copy at a good time.

Thank you for the educational comments (as usual), although I do worry that offloading is still most useful at 1 GPU. IIRC @wanchaol had been working on single GPU distributed that might be relevant for this work

**Comment by msaroufim** [comment]
@awgu so last week I did try to see if torch.compile vectorizes and uses OMP pragmas when compiling ADAM on cpu and it indeed does https://github.com/msaroufim/tinyoptimizer/blob/master/compiled_output_code/graph1.py#L31-L37

I was checking to see if I could codegen the kernel with torch.compile and rip it out so we could also use the code in eager, I won't do that since the code is kinda hard to read but at least to unblock I'm wondering whether you can just torch.compile the `opt.step()` to unblock for now

**Comment by awgu** [comment]
To-do: Make a pinned memory option (for parameters/gradients/both?)

File: torch/distributed/_composable/fsdp/_fsdp_api.py
Comment (gchanan) [inline torch/distributed/_composable/fsdp/_fsdp_api.py:68]: Naive question: why is this all or nothing, i.e. why doesn't the user have control over parameters, gradients, optimizer states separately

Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_api.py:68]: In my view, the main answer is that it is the simplest schedule that also happens to minimize the H2D/D2H copy bandwidth usage.

However, given that currently the optimizer step computation itself is the bottleneck and not the copying, I think it is fair to consider other schedules.

In this PR, FSDP is only responsible for (1) keeping sharded parameters on CPU, (2) copying them H2D as needed for forward/backward computation, and (3) copying sharded gradients D2H after reduce-scatter in backward. FSDP does not manipulate the optimizer explicitly.

Therefore, from FSDP's perspective, there are only two cases to discuss:
1. Keep sharded parameters on GPU but offload sharded gradients D2H to CPU
2. Keep sharded parameters on CPU but keep sharded gradients on GPU

Anything related to the optimizer state can be handled in the optimizer outside of FSDP. Note also that these two cases are only relevant to trying to run `optim.step()` on GPU, not CPU. If `optim.step()` is running on CPU, then the existing schedule is already optimal.

Some of my concerns with these two cases:
- Both of these cases would require the user to externally add logic such that parameters, gradients, and optimizer states are all on the same device at the time of `optim.step()` (i.e. this would not work out of the box with core optimizers). For example for case 1, the optimizer would need to copy a sharded gradient back to GPU, run the optimizer step for the parameter, and offload that gradient back to CPU. This could possibly be handled with optimizer hooks, though I wonder who is responsible for defining and registering these hooks. The efficiency for this kind of pattern should be evaluated too (e.g. for all parameters: copy H2D -> small computation -> copy D2H).
- The optimizer states contribute 2x memory compared to parameters or gradients each. For memory savings, we prefer for the optimizer states to be on CPU. However, we said that these two cases are only relevant when we run optimizer on _GPU_. This means either less memory savings or extra logic to also copy H2D/D2H optimizer states around `optim.step()`.
- PyTorch today does not let you have a tensor that is on a different device than its `.grad`. This means to support either of the two cases, we need to do something "extra" to save the parameters and gradients, while exposing them still to the user.
- Since FSDP is module-level, e.g. applied to each transformer block, this CPU offloading can be applied to one transformer block but not another (for example). This still offers some level of tuning in terms of how much they want to offload (albeit, not in the same way as the two aforementioned cases).

File: test/distributed/_composable/fsdp/test_fully_shard_training.py
Comment (weifengpy) [inline test/distributed/_composable/fsdp/test_fully_shard_training.py:319]: personally I misinterpreted ``CPUOffloadPolicy(False)`` as no cpu offloading, but actually it's ``pin_memory=False``. any chance this can be misinterpreated by others? 

flat-param FSDP use ``CPUOffload(False)`` to present ``offload_params=False``. this is probably why I was misinterpreting

Comment (awgu) [reply test/distributed/_composable/fsdp/test_fully_shard_training.py:319]: That is a good point. This difference might be confusing 🤔 

To explain some of my thoughts:
- For `mp_policy: MixedPrecisionPolicy`, since each of the dataclass's fields are optional, I did not want to have both `mp_policy=None` _and_ `mp_policy=MixedPrecisionPolicy(param_dtype=None, reduce_dtype=None)`, so I preferred just having `mp_policy=MixedPrecisionPolicy()` as default (no option for `None`).
- For `offload_policy`, now I am wondering whether it might be better to just have `offload_policy: Optional[OffloadPolicy] = None` (i.e. have it default to `None`). It does seem confusing as well that `OffloadPolicy` (base class) actually means no offloading.
- I am not sure how to make the fact that the `bool` arg to `CPUOffloadPolicy` is now changed to `pin_memory: bool` though.

File: test/distributed/_composable/fsdp/test_fully_shard_memory.py
Comment (Skylion007) [reply test/distributed/_composable/fsdp/test_fully_shard_memory.py:135]: Just some constant so it can be changed easily in the future if we need to for precision reasons (instead of a magic number).

File: test/distributed/_composable/fsdp/test_fully_shard_training.py
Comment (Skylion007) [reply test/distributed/_composable/fsdp/test_fully_shard_training.py:319]: Can we not treat it as a kwarg where we use it in the code / documentation. Some style guides suggest only using Bools as kwarg for this reason to help avoid confusion.

File: test/distributed/_composable/fsdp/test_fully_shard_memory.py
Comment (Skylion007) [inline test/distributed/_composable/fsdp/test_fully_shard_memory.py:38]: Do we not have a utility to note that we skipped the test like `pytest.skip`?

Comment (awgu) [reply test/distributed/_composable/fsdp/test_fully_shard_memory.py:38]: I think the problem is that this is using subtests. Do you know if there is such a utility to indicate a subtest is skipped?

(`self.run_subtests` is syntactic sugar for doing the product over configs and running subtests.)

File: test/distributed/_composable/fsdp/test_fully_shard_memory.py
Comment (awgu) [reply test/distributed/_composable/fsdp/test_fully_shard_memory.py:135]: I think the `1e6` is for converting to MB rather than being an epsilon in the numerical stability sense.

File: test/distributed/_composable/fsdp/test_fully_shard_training.py
Comment (awgu) [reply test/distributed/_composable/fsdp/test_fully_shard_training.py:319]: @Skylion007 Makes sense to me!

**Review by weifengpy — APPROVED** [review]
thanks, ``CPUOffloadPolicy(pin_memory=True)`` (kwargs) helps

**Comment by awgu** [comment]
Follow-ups for performance:
- Land `DTensor` support for fused Adam(W) so that we can leverage the new CPU fused Adam(W) kernels

**Pull request closed without merging**
